### PR TITLE
Git hooks

### PR DESCRIPTION
**Pull request by JonathanWoollett-Light**
# Reason for This PR

At the moment a contributor may need to wait ~30 minutes for the CI to run to be informed they have a trailing white space. By implementing a pre-commit git hook which both fully checks and rectifies formatting issues this considerable annoyance and time drain can be avoided.

## Description of Changes

1. Implemented a pre-commit git hook for formatting Rust and Python code
2. Adopted the [black](https://github.com/psf/black) python formatter.
3. Included additional `rustfmt` configuration options.

## Additional information

At the moment checks involving `pylint` and `cargo clippy` are absent as they can only be carried out in environments in which Firecracker can be built. These would for instance prevent making a commit editing a readme.md from windows as they would fail as Firecracker could not be built.

It is also worth noting all these configuration options are considered `unstable` please see https://github.com/firecracker-microvm/firecracker/pull/3037#discussion_r904763529 for why this is acceptable. 

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The issue which led to this PR has a clear conclusion.
- [x] This PR follows the solution outlined in the related issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes follow the [Runbook for Firecracker API changes][2].
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

[1]: https://github.com/rust-vmm
[2]: ../docs/api-change-runbook.md

### CODE REVIEW

File: tests/fmt.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+comment_width=100
+wrap_comments=true
+format_code_in_doc_comments=true
+format_strings=true
+license_template_path="./tests/license.txt"
```
[the remark this answers is not in the document]
Comment (JonathanWoollett-Light) [reply tests/fmt.toml:5]: > This is very cool, but looks like it's deprecated?

Like most of these options rather it is unstable, you will not see it under [v1.5.0](https://rust-lang.github.io/rustfmt/?version=v1.5.0) but you will see it under [v2.0.0-rc.2](https://rust-lang.github.io/rustfmt/?version=v2.0.0-rc.2).

> Also it overlaps with test_licenses.py

I noticed this in development. `test_licenses.py` I believe requires a considerable rework, we could use the regex given in `license.txt` and apply it to all files (likely reducing `test_licenses.py` to <50 lines) and removing `license_template_path`.

At the time of development I considered this to be a reasonably small but separate next step in refining this testing. A considerable factor here is that utilizing pytest is a considerable dependency to the git hook and could potentially disabled git interaction in many developers current environments.

At the moment I believe this approach is the simplest yet still effective implementation and that it creates more somewhat atomic tasks we could work on should we desired to continue work in the area.

File: pre-commit [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    extension="${filename##*.}"
+    if [ "$extension" = "rs" ]; then
+        # Read rustfmt config, replace '\n' with ','
+        rustfmt_config_1="$(sed -z "s/\n/,/g;s/,$/\n/" /tests/fmt.toml)"
+        # Remove '"'
+        rustfmt_config_2="${rustfmt_config_1//\"}"
+        # We first do a check run, this will fail when it finds a non-matching license.
+        rustfmt $i --check --config $rustfmt_config_2
+        # Run `cargo fmt` for this file
+        rustfmt $i --config $rustfmt_config_2
```
[the remark this answers is not in the document]
Comment (JonathanWoollett-Light) [reply pre-commit:32]: Yes but, see https://github.com/firecracker-microvm/firecracker/pull/3037#discussion_r904763529

**Review by pb8o — COMMENTED** [review]
Could we put the commits that just change style in .git-blame-ignore-revs?
Or maybe mark the commit with [style] so we can easily grep it later?https://github.com/github-community/community/discussions/5033

**Comment by JonathanWoollett-Light** [comment]
@pb8o I've added `.git-blame-ignore-revs` this does assume/require the individual viewing has set their configuration with `git config --global blame.ignoreRevsFile .git-blame-ignore-revs`

File: .git-blame-ignore-revs [hunk 3]
```diff
@@ -0,0 +1,2 @@
+b637c0d4cfdeb4c9c924405076b6dbdc687920b5
+22833835519c163113058e71458a4bfdce8e1470
```
Comment (pb8o) [inline .git-blame-ignore-revs:2]: newline missing

**Review by pb8o — APPROVED** [review]

**Review by pb8o — APPROVED** [review]

File: tests/integration_tests/build/test_pylint.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    # List of linter commands that should be executed for each file
+    linter_cmd = (
+        # Pylint
+        "python3 -m pylint --jobs=0 --persistent=no --score=no "
+        '--output-format=colorized --attr-rgx="[a-z_][a-z0-9_]{1,30}$" '
+        '--argument-rgx="[a-z_][a-z0-9_]{1,35}$" '
+        '--variable-rgx="[a-z_][a-z0-9_]{1,30}$" --disable='
+        "fixme,too-many-instance-attributes,import-error,"
+        "too-many-locals,too-many-arguments,consider-using-f-string,"
+        "consider-using-with,implicit-str-concat"
```
Comment (bchalios) [inline tests/integration_tests/build/test_pylint.py:23]: nit: can you indent lines after the first?

**Review by bchalios — DISMISSED** [review]
LGTM

File: tests/integration_tests/build/test_pylint.py (hunk 4 above)
Comment (JonathanWoollett-Light) [reply tests/integration_tests/build/test_pylint.py:23]: I'm not sure what you mean here could you clarify?

**Review by bchalios — DISMISSED** [review]

**Review by pb8o — DISMISSED** [review]

**Review by bchalios — APPROVED** [review]

**Review by pb8o — APPROVED** [review]

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -57,7 +57,15 @@ If you just want to receive feedback for a contribution proposal, open an “RFC
 ## Contribution Quality Standards
 
 Most quality and style standards are enforced automatically during integration
-testing. Your contribution needs to meet the following standards:
+testing. For ease of use you can setup a git pre-commit hook by running the
+following in the Firecracker root directory:
+
+```
+cargo install rusty-hooks
```
[the remark this answers is not in the document]
Comment (JonathanWoollett-Light) [reply CONTRIBUTING.md:64]: Just posted https://github.com/firecracker-microvm/firecracker/pull/3048 to fix this, thank you for pointing it out.

File: rusty-hook.toml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# To set up git hooks run
+# ```
+# cargo install rusty-hooks
```
[the remark this answers is not in the document]
Comment (JonathanWoollett-Light) [reply rusty-hook.toml:3]: See https://github.com/firecracker-microvm/firecracker/pull/3037#discussion_r915700851.

**Pull request merged**